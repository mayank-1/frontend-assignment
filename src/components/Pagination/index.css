.container {
    display: flex;
    width: 100%;
    justify-content: center;
    align-items: center;
    margin-top: 20px;
    margin-bottom: 20px;
    flex-wrap: wrap;
    gap: 10px;

    .ellipsis {
        align-self: baseline;
        color: #6f7373;
    }

    .total-pages {
        color: #6f7373;
        font-size: 12px;
    }
}