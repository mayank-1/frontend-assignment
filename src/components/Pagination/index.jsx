import React from "react";
import PropTypes from "prop-types";

// COMPONENTS
import Button from "./../Button";

// CSS
import "./index.css";

const Pagination = ({ currentPage, totalPages, onPageChange }) => {
  const isLastFourPages = currentPage > totalPages - 4;

  const getPageNumbers = () => {
    if (!totalPages || totalPages <= 0) return [];

    const pageNumbers = [];

    // Determine the range of page numbers to display
    if (isLastFourPages) {
      // Show the last four pages
      for (let i = Math.max(1, totalPages - 3); i <= totalPages; i++) {
        pageNumbers.push(i);
      }
    } else {
      // Show 3 pages starting from the current page
      const startPage = currentPage;
      const endPage = Math.min(currentPage + 2, totalPages - 1);

      for (let i = startPage; i <= endPage; i++) {
        pageNumbers.push(i);
      }
    }

    return pageNumbers;
  };

  if (!totalPages || totalPages <= 0) return null;

  const showFirstPageButton = currentPage > 4;

  return (
    <div className="container" role="navigation" aria-label="Pagination">
      <Button
        onClick={() => onPageChange(currentPage - 1)}
        disabled={currentPage === 1}
        label="Previous page"
      >
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="8"
          height="12"
          viewBox="0 0 8 12"
          focusable="false"
          aria-hidden="true"
        >
          <path d="M5.874.35a1.28 1.28 0 011.761 0 1.165 1.165 0 010 1.695L3.522 6l4.113 3.955a1.165 1.165 0 010 1.694 1.28 1.28 0 01-1.76 0L0 6 5.874.35z"></path>
        </svg>
        <span>Previous</span>
      </Button>

      {showFirstPageButton && (
        <>
          <Button
            onClick={() => onPageChange(1)}
            active={currentPage === 1}
            label="Page 1"
          >
            1
          </Button>
          <span className="ellipsis">...</span>
        </>
      )}

      {getPageNumbers().map((number) => (
        <Button
          key={number}
          onClick={() => onPageChange(number)}
          active={currentPage === number}
          label={`Page ${number}`}
        >
          {number}
        </Button>
      ))}

      {!isLastFourPages && totalPages > 4 && (
        <>
          <span className="ellipsis">...</span>
          <span className="total-pages">{totalPages}</span>
        </>
      )}

      <Button
        onClick={() => onPageChange(currentPage + 1)}
        disabled={currentPage === totalPages}
        label="Next page"
      >
        <span>Next</span>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="8"
          height="12"
          viewBox="0 0 8 12"
          focusable="false"
          aria-hidden="true"
        >
          <path d="M2.126.35a1.28 1.28 0 00-1.761 0 1.165 1.165 0 000 1.695L4.478 6 .365 9.955a1.165 1.165 0 000 1.694 1.28 1.28 0 001.76 0L8 6 2.126.35z"></path>
        </svg>
      </Button>
    </div>
  );
};

Pagination.propTypes = {
  currentPage: PropTypes.number.isRequired,
  totalPages: PropTypes.number.isRequired,
  onPageChange: PropTypes.func.isRequired,
};

export default Pagination;
