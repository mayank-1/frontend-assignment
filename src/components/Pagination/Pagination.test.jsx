import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, it, expect } from "vitest";
import Pagination from "./index";

describe("Pagination", () => {
  it("renders correctly with basic props", () => {
    render(
      <Pagination currentPage={2} totalPages={5} onPageChange={() => {}} />
    );
    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  it('renders disabled "Previous" button on first page', () => {
    render(
      <Pagination currentPage={1} totalPages={5} onPageChange={() => {}} />
    );
    expect(screen.getByRole("button", { name: /Previous/i })).toBeDisabled();
    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  it('renders "Previous" button on later pages', () => {
    render(
      <Pagination currentPage={3} totalPages={5} onPageChange={() => {}} />
    );
    expect(
      screen.getByRole("button", { name: /Previous/i })
    ).not.toBeDisabled();
    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  it('renders "Next" button on earlier pages', () => {
    render(
      <Pagination currentPage={2} totalPages={5} onPageChange={() => {}} />
    );
    expect(screen.getByRole("button", { name: /Next/i })).not.toBeDisabled();
    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  it('renders disabled "Next" button on last page', () => {
    render(
      <Pagination currentPage={5} totalPages={5} onPageChange={() => {}} />
    );
    expect(screen.getByRole("button", { name: /Next/i })).toBeDisabled();
    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  it("renders page numbers correctly", () => {
    render(
      <Pagination currentPage={3} totalPages={7} onPageChange={() => {}} />
    );
    const pageButtons = screen.getAllByRole("button", { name: /page/i });
    expect(pageButtons.length).toBe(5);
    expect(pageButtons[0].textContent).toBe("Previous");
    expect(pageButtons[1].textContent).toBe("3");
    expect(pageButtons[2].textContent).toBe("4");
    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  it("renders ellipsis for missing pages in the middle", () => {
    render(
      <Pagination currentPage={4} totalPages={10} onPageChange={() => {}} />
    );
    expect(screen.getByText("...")).toBeInTheDocument();
    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  it("calls onPageChange on button click", async () => {
    const onPageChange = vi.fn();
    render(
      <Pagination currentPage={2} totalPages={5} onPageChange={onPageChange} />
    );

    const nextButton = screen.getByRole("button", { name: /Next/i });
    await userEvent.click(nextButton);

    expect(onPageChange).toHaveBeenCalledTimes(1);
    expect(onPageChange).toHaveBeenCalledWith(3);
  });

  it("does not call onPageChange on disabled button click", async () => {
    const onPageChange = vi.fn();
    render(
      <Pagination currentPage={1} totalPages={5} onPageChange={onPageChange} />
    );

    const previousButton = screen.getByRole("button", { name: /Previous/i });
    await userEvent.click(previousButton);

    expect(onPageChange).not.toHaveBeenCalled();
  });
});
