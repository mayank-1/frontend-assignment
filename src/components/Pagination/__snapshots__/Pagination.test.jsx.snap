// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > renders "Next" button on earlier pages 1`] = `
<div
  aria-label="Pagination"
  class="container"
  role="navigation"
>
  <button
    aria-label="Previous page"
    class=""
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5.874.35a1.28 1.28 0 011.761 0 1.165 1.165 0 010 1.695L3.522 6l4.113 3.955a1.165 1.165 0 010 1.694 1.28 1.28 0 01-1.76 0L0 6 5.874.35z"
      />
    </svg>
    <span>
      Previous
    </span>
  </button>
  <button
    aria-current="page"
    aria-label="Page 2"
    class="active"
  >
    2
  </button>
  <button
    aria-label="Page 3"
    class=""
  >
    3
  </button>
  <button
    aria-label="Page 4"
    class=""
  >
    4
  </button>
  <button
    aria-label="Page 5"
    class=""
  >
    5
  </button>
  <button
    aria-label="Next page"
    class=""
  >
    <span>
      Next
    </span>
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.126.35a1.28 1.28 0 00-1.761 0 1.165 1.165 0 000 1.695L4.478 6 .365 9.955a1.165 1.165 0 000 1.694 1.28 1.28 0 001.76 0L8 6 2.126.35z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination > renders "Previous" button on later pages 1`] = `
<div
  aria-label="Pagination"
  class="container"
  role="navigation"
>
  <button
    aria-label="Previous page"
    class=""
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5.874.35a1.28 1.28 0 011.761 0 1.165 1.165 0 010 1.695L3.522 6l4.113 3.955a1.165 1.165 0 010 1.694 1.28 1.28 0 01-1.76 0L0 6 5.874.35z"
      />
    </svg>
    <span>
      Previous
    </span>
  </button>
  <button
    aria-label="Page 2"
    class=""
  >
    2
  </button>
  <button
    aria-current="page"
    aria-label="Page 3"
    class="active"
  >
    3
  </button>
  <button
    aria-label="Page 4"
    class=""
  >
    4
  </button>
  <button
    aria-label="Page 5"
    class=""
  >
    5
  </button>
  <button
    aria-label="Next page"
    class=""
  >
    <span>
      Next
    </span>
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.126.35a1.28 1.28 0 00-1.761 0 1.165 1.165 0 000 1.695L4.478 6 .365 9.955a1.165 1.165 0 000 1.694 1.28 1.28 0 001.76 0L8 6 2.126.35z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination > renders correctly with basic props 1`] = `
<div
  aria-label="Pagination"
  class="container"
  role="navigation"
>
  <button
    aria-label="Previous page"
    class=""
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5.874.35a1.28 1.28 0 011.761 0 1.165 1.165 0 010 1.695L3.522 6l4.113 3.955a1.165 1.165 0 010 1.694 1.28 1.28 0 01-1.76 0L0 6 5.874.35z"
      />
    </svg>
    <span>
      Previous
    </span>
  </button>
  <button
    aria-current="page"
    aria-label="Page 2"
    class="active"
  >
    2
  </button>
  <button
    aria-label="Page 3"
    class=""
  >
    3
  </button>
  <button
    aria-label="Page 4"
    class=""
  >
    4
  </button>
  <button
    aria-label="Page 5"
    class=""
  >
    5
  </button>
  <button
    aria-label="Next page"
    class=""
  >
    <span>
      Next
    </span>
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.126.35a1.28 1.28 0 00-1.761 0 1.165 1.165 0 000 1.695L4.478 6 .365 9.955a1.165 1.165 0 000 1.694 1.28 1.28 0 001.76 0L8 6 2.126.35z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination > renders disabled "Next" button on last page 1`] = `
<div
  aria-label="Pagination"
  class="container"
  role="navigation"
>
  <button
    aria-label="Previous page"
    class=""
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5.874.35a1.28 1.28 0 011.761 0 1.165 1.165 0 010 1.695L3.522 6l4.113 3.955a1.165 1.165 0 010 1.694 1.28 1.28 0 01-1.76 0L0 6 5.874.35z"
      />
    </svg>
    <span>
      Previous
    </span>
  </button>
  <button
    aria-label="Page 1"
    class=""
  >
    1
  </button>
  <span
    class="ellipsis"
  >
    ...
  </span>
  <button
    aria-label="Page 2"
    class=""
  >
    2
  </button>
  <button
    aria-label="Page 3"
    class=""
  >
    3
  </button>
  <button
    aria-label="Page 4"
    class=""
  >
    4
  </button>
  <button
    aria-current="page"
    aria-label="Page 5"
    class="active"
  >
    5
  </button>
  <button
    aria-label="Next page"
    class=""
    disabled=""
  >
    <span>
      Next
    </span>
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.126.35a1.28 1.28 0 00-1.761 0 1.165 1.165 0 000 1.695L4.478 6 .365 9.955a1.165 1.165 0 000 1.694 1.28 1.28 0 001.76 0L8 6 2.126.35z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination > renders disabled "Previous" button on first page 1`] = `
<div
  aria-label="Pagination"
  class="container"
  role="navigation"
>
  <button
    aria-label="Previous page"
    class=""
    disabled=""
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5.874.35a1.28 1.28 0 011.761 0 1.165 1.165 0 010 1.695L3.522 6l4.113 3.955a1.165 1.165 0 010 1.694 1.28 1.28 0 01-1.76 0L0 6 5.874.35z"
      />
    </svg>
    <span>
      Previous
    </span>
  </button>
  <button
    aria-current="page"
    aria-label="Page 1"
    class="active"
  >
    1
  </button>
  <button
    aria-label="Page 2"
    class=""
  >
    2
  </button>
  <button
    aria-label="Page 3"
    class=""
  >
    3
  </button>
  <span
    class="ellipsis"
  >
    ...
  </span>
  <span
    class="total-pages"
  >
    5
  </span>
  <button
    aria-label="Next page"
    class=""
  >
    <span>
      Next
    </span>
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.126.35a1.28 1.28 0 00-1.761 0 1.165 1.165 0 000 1.695L4.478 6 .365 9.955a1.165 1.165 0 000 1.694 1.28 1.28 0 001.76 0L8 6 2.126.35z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination > renders ellipsis for missing pages in the middle 1`] = `
<div
  aria-label="Pagination"
  class="container"
  role="navigation"
>
  <button
    aria-label="Previous page"
    class=""
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5.874.35a1.28 1.28 0 011.761 0 1.165 1.165 0 010 1.695L3.522 6l4.113 3.955a1.165 1.165 0 010 1.694 1.28 1.28 0 01-1.76 0L0 6 5.874.35z"
      />
    </svg>
    <span>
      Previous
    </span>
  </button>
  <button
    aria-current="page"
    aria-label="Page 4"
    class="active"
  >
    4
  </button>
  <button
    aria-label="Page 5"
    class=""
  >
    5
  </button>
  <button
    aria-label="Page 6"
    class=""
  >
    6
  </button>
  <span
    class="ellipsis"
  >
    ...
  </span>
  <span
    class="total-pages"
  >
    10
  </span>
  <button
    aria-label="Next page"
    class=""
  >
    <span>
      Next
    </span>
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.126.35a1.28 1.28 0 00-1.761 0 1.165 1.165 0 000 1.695L4.478 6 .365 9.955a1.165 1.165 0 000 1.694 1.28 1.28 0 001.76 0L8 6 2.126.35z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination > renders page numbers correctly 1`] = `
<div
  aria-label="Pagination"
  class="container"
  role="navigation"
>
  <button
    aria-label="Previous page"
    class=""
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5.874.35a1.28 1.28 0 011.761 0 1.165 1.165 0 010 1.695L3.522 6l4.113 3.955a1.165 1.165 0 010 1.694 1.28 1.28 0 01-1.76 0L0 6 5.874.35z"
      />
    </svg>
    <span>
      Previous
    </span>
  </button>
  <button
    aria-current="page"
    aria-label="Page 3"
    class="active"
  >
    3
  </button>
  <button
    aria-label="Page 4"
    class=""
  >
    4
  </button>
  <button
    aria-label="Page 5"
    class=""
  >
    5
  </button>
  <span
    class="ellipsis"
  >
    ...
  </span>
  <span
    class="total-pages"
  >
    7
  </span>
  <button
    aria-label="Next page"
    class=""
  >
    <span>
      Next
    </span>
    <svg
      aria-hidden="true"
      focusable="false"
      height="12"
      viewBox="0 0 8 12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.126.35a1.28 1.28 0 00-1.761 0 1.165 1.165 0 000 1.695L4.478 6 .365 9.955a1.165 1.165 0 000 1.694 1.28 1.28 0 001.76 0L8 6 2.126.35z"
      />
    </svg>
  </button>
</div>
`;
